runtime/caml/platform.h

```diff
@@ -120,6 +92,231 @@ void caml_plat_broadcast(caml_plat_cond*);
+ * |-----------|---------------+---------------------|
+ * | Reset     | `reset`       | automatic at `flip` |
+ * | Arrive    | `arrive`      | `arrive`            |
+ * | Check     | `is_released` | `sense_has_flipped` |
+ * | Block     | `wait`        | `wait_sense`        |
+ * | Release   | `release`     | `flip`              |
+ *
+ * The lifecycle is as follows:
+ *
+ *      Reset (1 thread)          (other threads)
+ *              |                       |
+ *              +----------+------------+
+ *                         |
+ *                      Arrive (all threads)
+ *                         |
+ *                         | check arrival number
+ *              +----------+------------+
+ *              |                       |
+ *       Check or Block              Release
+ *     (non-final threads)         (final thread)
+ *              |                       |
+ */
+typedef struct caml_plat_barrier {
+  caml_plat_futex futex;
+  atomic_uintnat arrived; /* includes sense bit */
+} caml_plat_barrier;
+#define CAML_PLAT_BARRIER_INITIALIZER \
+  { CAML_PLAT_FUTEX_INITIALIZER(0), ATOMIC_UINTNAT_INIT(0) }
+
+typedef uintnat barrier_status;
+/* Arrive at the barrier, returns the number of parties that have
+   arrived at the barrier (including this one); the caller should
+   check whether it is the last expected party to arrive, and release
+   or flip the barrier if so.
+
+   In a sense-reversing barrier, this also encodes the current sense
+   of the barrier in BARRIER_SENSE_BIT, which should be masked off if
+   checking for the last arrival. */
+Caml_inline barrier_status caml_plat_barrier_arrive(caml_plat_barrier* barrier)
+{
+  return 1 + atomic_fetch_add(&barrier->arrived, 1);
+}
+#define BARRIER_SENSE_BIT 0x100000
+
+/* -- Single-sense --
+
+   Futex states:
+   - 0 if released
+   - 1 if nobody is blocking (but they may be spinning)
+   - 2 if anybody is blocking (or about to)
+ */
+#define Barrier_released 0
+#define Barrier_unreleased 1
+#define Barrier_contested 2
+
+/* Reset the barrier to 0 arrivals, block new waiters */
+Caml_inline void caml_plat_barrier_reset(caml_plat_barrier* barrier) {
+  atomic_store_relaxed(&barrier->futex.value, Barrier_unreleased);
+  /* threads check arrivals before the futex, 'release' ordering
+     ensures they see it reset */
+  atomic_store_release(&barrier->arrived, 0);
+}
+/* Check if the barrier has been released */
+Caml_inline int caml_plat_barrier_is_released(caml_plat_barrier* barrier) {
+  return atomic_load_acquire(&barrier->futex.value) == Barrier_released;
+}
+
+/* Release and wait, but on a futex only.
+
+   This is like a(n inverted) binary semaphore, but with no decrement
+   on `wait`. That is, `release` sets the futex to 0, which `wait`
+   waits for.
+
+   A futex used this way is 0 (Barrier_released) when released, and
+   nonzero otherwise. It should be set to 1 (Barrier_unreleased) to
+   block.
+*/
+void caml_plat_barrier_raw_release(caml_plat_futex* futex);
```

Comment:
Looking for a name, the only existing similar synchronisation object I can find is Windows' "manual reset events" ([Win32 API](https://learn.microsoft.com/en-us/windows/win32/sync/event-objects), [.NET class](https://learn.microsoft.com/en-us/dotnet/api/system.threading.manualresetevent?view=net-7.0)), which I'm not a huge fan of. My best idea is `binary_latch`, where (non-binary) latches exist as [`std::latch` in C++](https://en.cppreference.com/w/cpp/thread/latch) and as [`CountDownLatch` in Java](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/concurrent/CountDownLatch.html). It might be worth extracting the operation to a new type entirely, which could also help clarify and abstract uses of the `Barrier_*` constants.